### PR TITLE
Allow avatar customization

### LINKED ISSUE #253: Allow injecting custom avatars
Allow injecting custom avatars as it's a common thing for 3rd parties to have their own avatar.

### PR DESCRIPTION
**Pull request by pinarol**
Closes https://github.com/Automattic/Gravatar-SDK-iOS/issues/253

### Description

Introduces `AvatarType` to allow 3rd parties inject their own avatar. There are 3 levels to it:

1. Avatar is a `UIImageView` or a subclass of it. We apply its style normally unless they set `skipStyling = true`.
2. Avatar is a `UIView` that has a `UIImageView` in it. In this case we still take care of downloading and setting the image but we don't apply any style because it's unclear where exactly to apply the style, to the parent or to the `UIImageView`?. It's best to let this be determined by the 3rd party component.
3. Fully custom. Avatar is provided via "AvatarProviding" protocol.

### Testing Steps

CI green.

### CODE REVIEW

File: Sources/GravatarUI/ProfileView/Avatar/AvatarType.swift [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            provider
+        }
+    }
+}
+
+/// Defines a UIView that has a UIImageView subview.
+@MainActor
+public protocol ImageViewWrapper {
+    // The `UIImageView` to display the avatar.
+    var imageView: UIImageView { get }
```
Comment (etoledom) [inline Sources/GravatarUI/ProfileView/Avatar/AvatarType.swift]:
```suggestion
    /// The `UIImageView` to display the avatar.
    var imageView: UIImageView { get }
```

File: Sources/GravatarUI/ProfileView/Avatar/AvatarType.swift [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+/// Provides a UIView to be used as the avatar in a `BaseProfileView`. This is for taking full control over the avatar..
+@MainActor
+public protocol AvatarProviding {
+    /// `UIView` to insert into the avatar's slot in a `BaseProfileView`.
+    var avatarView: UIView { get }
+
+    /// `BaseProfileView` calls this method to load the avatar from a URL.
+    /// You may not need to implement this method if the image is downloaded else where.
+    func setImage(with source: URL?, placeholder: UIImage?, options: [ImageSettingOption]?, completion: ((Bool) -> Void)?)
+
```
Comment (etoledom) [inline Sources/GravatarUI/ProfileView/Avatar/AvatarType.swift:79]: I thought for a second about making this async, but then I noticed that we don't have an async alternative for the UIImageVIew.gravatar extension 🤔 

Maybe we can do this at some point in the future

**Review by etoledom — APPROVED** [review]
Looking great!

File: Sources/GravatarUI/ProfileView/Avatar/AvatarType.swift (hunk 2 above)
Comment (pinarol) [reply Sources/GravatarUI/ProfileView/Avatar/AvatarType.swift:79]: Yep, nowadays I'll work on that.

File: Sources/GravatarUI/ProfileView/Avatar/AvatarType.swift [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+/// Defines the different types of avatars that can be passed to the `BaseProfileView` subclasses.
+public enum AvatarType {
+    /// Avatar is a `UIImageView` or a subclass of it.
+    /// If `skipStyling` is `true` then no size & shape & border changes will be applied to the `UIImageView` so the appearance needs to be managed externally.
+    case imageView(UIImageView, skipStyling: Bool = false)
+
+    /// Avatar is a `UIView` that has a `UIImageView` and its appearance needs to be managed externally.
+    case imageViewWrapper(ImageViewWrapper)
+
+    /// Avatar is a custom `UIView` whose appearance is fully managed externally.
```
Comment (pinarol) [inline Sources/GravatarUI/ProfileView/Avatar/AvatarType.swift]:
```suggestion
    /// Avatar is a custom `UIView` whose appearance and contents are fully managed externally.
```

File: Sources/GravatarUI/ProfileView/Avatar/AvatarType.swift [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+@MainActor
+public protocol ImageViewWrapper {
+    /// The `UIImageView` to display the avatar.
+    var imageView: UIImageView { get }
+
+    /// The parent view for the imageView. It doesn't have to be the direct parent.
+    var baseView: UIView { get }
+}
+
+/// Provides a UIView to be used as the avatar in a `BaseProfileView`. This is for taking full control over the avatar..
```
Comment (pinarol) [inline Sources/GravatarUI/ProfileView/Avatar/AvatarType.swift]:
```suggestion
/// Provides a UIView to be used as the avatar in a `BaseProfileView`. This is for taking full control over the avatar.
```

File: Sources/GravatarUI/ProfileView/Avatar/AvatarType.swift (hunk 2 above)
Comment (etoledom) [reply Sources/GravatarUI/ProfileView/Avatar/AvatarType.swift:79]: Oh right, there's an old ticket for this already https://github.com/Automattic/Gravatar-SDK-iOS/issues/164
Nice! 🙏

**Pull request merged**